# Comparaison de la prédiction avec ou sans la feature "ENERGYSTARScore"

1. Sans "ENERGYSTARScore"
2. Avec "ENERGYSTARScore"

In [12]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [13]:
pd.set_option("display.max_columns", None)
datas = pd.read_csv("Pélec_14_gaz_train.csv")
relevant_features = pd.read_csv("Pélec_16_relevant_features.csv")

datas.head()

OSEBuildingID PrimaryPropertyType  CouncilDistrictCode  YearBuilt  \
0            1.0               hotel                  7.0     1927.0   
1            2.0               hotel                  7.0     1996.0   
2            3.0               hotel                  7.0     1969.0   
3            5.0               hotel                  7.0     1926.0   
4            8.0               hotel                  7.0     1980.0   

   NumberofBuildings  NumberofFloors  PropertyGFATotal  PropertyGFAParking  \
0                1.0            12.0           88434.0                 0.0   
1                1.0            11.0          103566.0             15064.0   
2                1.0            41.0          961990.0                 0.0   
3                1.0            10.0           61320.0                 0.0   
4                1.0            18.0          119890.0             12460.0   

   PropertyGFABuildings  ENERGYSTARScore  SiteEUIkBtusf  SiteEUIWNkBtusf  \
0               88434.0             65.0           78.9             80.3   
1               88502.0             51.0           94.4             99.0   
2              961990.0             18.0           96.6             99.7   
3               61320.0              1.0          460.4            462.5   
4              107430.0             67.0          120.1            122.1   

   SourceEUIkBtusf  SourceEUIWNkBtusf  SiteEnergyUsekBtu  SiteEnergyUseWNkBtu  \
0            173.5              175.1          6981428.0            7097539.0   
1            191.3              195.2          8354235.0            8765788.0   
2            242.7              246.5         73130656.0           75506272.0   
3            636.3              643.2         28229320.0           28363444.0   
4            228.8              227.1         14829099.0           15078243.0   

   SteamUsekBtu  ElectricitykWh  ElectricitykBtu  NaturalGastherms  \
0     2023032.0       1080307.0        3686160.0           12724.0   
1           0.0       1144563.0        3905411.0           44490.0   
2    19660404.0      14583930.0       49762435.0           37099.0   
3    23458518.0        811521.0        2769023.0           20019.0   
4           0.0       1777841.0        6066245.0           87631.0   

   NaturalGaskBtu  OtherFuelUsekBtu       Outlier  TotalGHGEmissions  \
0       1272388.0               0.0           NaN             249.43   
1       4448985.0               0.0           NaN             263.51   
2       3709900.0               0.0           NaN            2061.48   
3       2001894.0               0.0  high outlier            1936.34   
4       8763105.0               0.0           NaN             507.70   

   GHGEmissionsIntensity  SiteEUIkBtusf_diff1516  SiteEUIWNkBtusf_diff1516  \
0                   2.64                2.799997                  4.000003   
1                   2.38                0.400003                  0.000000   
2                   1.92                0.000000                  0.000000   
3                  31.38                0.000000                  0.000000   
4                   4.02                0.000000                  0.000000   

   SourceEUIkBtusf_diff1516  SourceEUIWNkBtusf_diff1516  \
0                       9.0                        13.9   
1                       0.0                         0.0   
2                       0.0                         0.0   
3                       0.0                         0.0   
4                       0.0                         0.0   

   SiteEnergyUsekBtu_diff1516  SiteEnergyUseWNkBtu_diff1516  \
0                    244934.5                      475482.0   
1                     33698.0                      310244.0   
2                         0.0                      806456.0   
3                         0.0                           0.0   
4                         0.0                           0.0   

   SteamUsekBtu_diff1516  ElectricitykWh_diff1516  ElectricitykBtu_diff1516  \
0                    

## classement des variables

* Les variables nominatives seront transformées par OneHotEncoder
* Les variables ordinales seront transformées par OrdinalEncoder
* Les variables numériques seront normalisées par QuantileTransformer

In [14]:
nom_features = ["PrimaryPropertyType", "CouncilDistrictCode", "Outlier"
    ]
ord_features = ["YearBuilt", "NumberofFloors", "NumberofBuildings", "HasParking" 
    ]
num_features = ["PropertyGFABuildings", "PropertyGFAParking", 
    "GFAperFloor", "GFAperBuilding", 
    "percGFABuilding", "percGFAParking", 
    
    "SteamUsekBtu", "ElectricitykBtu", "NaturalGaskBtu", "OtherFuelUsekBtu", 
    "SteamUsekBtu_diff1516", "ElectricitykBtu_diff1516", "NaturalGaskBtu_diff1516", 

    "SteamperGFA", "ElecperGFA", "GasperGFA", "FuelperGFA", "percSteam", "percElec", "percGas", "percFuel", 

    "WNfactor", 
    "SteamWN", "ElecWN", "GasWN", "FuelWN", 
    "SteamWNperGFA", "ElecWNperGFA", "GasWNperGFA", "FuelWNperGFA", 
    "percSteamWN", "percElecWN", "percGasWN", "percFuelWN", 
    "WNfactor_diff", "Steam_diff1516WN", "Elec_diff1516WN", "Gas_diff1516WN"
    ]
add_feature = ["ENERGYSTARScore"]

cible = ["TotalGHGEmissions"]

In [15]:
def to_log(X):
  return np.log(X + 1)

def to_exp(X):
  return np.exp(X) - 1

In [16]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import QuantileTransformer, OneHotEncoder, OrdinalEncoder, FunctionTransformer, StandardScaler
from sklearn.compose import ColumnTransformer, TransformedTargetRegressor
from sklearn.kernel_ridge import KernelRidge
from sklearn.pipeline import Pipeline

## 1. Sans "ENERGYSTARScore"

In [17]:
X = datas[nom_features + ord_features + num_features]
y = datas[cible]

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=42)

pipe = Pipeline([("qt", FunctionTransformer(func= to_log, inverse_func= to_exp)), ("scaler", StandardScaler())])

ct = ColumnTransformer([("onehotencoder", OneHotEncoder(handle_unknown="ignore"), nom_features),
    ("quantiletransformer", pipe, num_features),
    ("ordinalencoder", OrdinalEncoder(handle_unknown= "use_encoded_value", unknown_value= -1), ord_features)
    ])

ct.fit(X_train)

X_train_ct = pd.DataFrame(ct.transform(X_train), columns= list(ct.transformers_[0][1].get_feature_names(nom_features)) + 
    num_features + ord_features)
X_test_ct = pd.DataFrame(ct.transform(X_test), columns= list(ct.transformers_[0][1].get_feature_names(nom_features)) + num_features
    + ord_features)

In [18]:
X_train_relevant = X_train_ct[relevant_features.loc[:, "columns"]]
X_test_relevant = X_test_ct[relevant_features.loc[:, "columns"]]
y_train = y_train.reset_index(drop= True)
y_test = y_test.reset_index(drop= True)


In [19]:
model = TransformedTargetRegressor(regressor=KernelRidge(alpha= 1, kernel= "poly", gamma= 0.1), 
    transformer=QuantileTransformer(output_distribution="normal"))

model.fit(X_train_relevant, y_train)

print(f"R2 sans 'EnergyStarScore': {model.score(X_test_relevant, y_test)}")

R2 sans 'EnergyStarScore': 0.9489332339987281


## 2. Avec "ENERGYSTARScore"

In [20]:
X = datas[nom_features + ord_features + num_features + add_feature]
y = datas[cible]

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=42)

ct = ColumnTransformer([("onehotencoder", OneHotEncoder(handle_unknown="ignore"), nom_features),
    ("quantiletransformer", FunctionTransformer(func= to_log), num_features),
    ("ordinalencoder", OrdinalEncoder(handle_unknown= "use_encoded_value", unknown_value= -1), ord_features + add_feature)
    ])
    
ct.fit(X_train)

X_train_ct = pd.DataFrame(ct.transform(X_train), columns= list(ct.transformers_[0][1].get_feature_names(nom_features)) + 
    num_features + ord_features + add_feature)
X_test_ct = pd.DataFrame(ct.transform(X_test), columns= list(ct.transformers_[0][1].get_feature_names(nom_features)) + num_features
    + ord_features + add_feature)

In [21]:
X_train_relevant = X_train_ct[relevant_features.loc[:, "columns"].tolist() + add_feature]
X_test_relevant = X_test_ct[relevant_features.loc[:, "columns"].tolist() + add_feature]
y_train = y_train.reset_index(drop= True)
y_test = y_test.reset_index(drop= True)


In [22]:
model = TransformedTargetRegressor(regressor=KernelRidge(alpha= 1, kernel= "poly", gamma= 0.1), 
    transformer=QuantileTransformer(output_distribution="normal"))

model.fit(X_train_relevant, y_train)

print(f"R2 avec 'EnergyStarScore': {model.score(X_test_relevant, y_test)}")

R2 avec 'EnergyStarScore': 0.9171713866598957


R2 sans 'EnergyStarScore': 0.9489332339987281

R2 avec 'EnergyStarScore': 0.9171713866598957

On peut constater un meilleur score lors de la prédiction sans utiliser l'ENERGYSTARScore. Cela est d'autant plus intéressant que ça évite un calcul compliqué supplémentaire